paddle/operators/beam_search_op.h

```diff
@@ -0,0 +1,148 @@
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License. */
+
+#pragma once
+
+#include "paddle/framework/lod_tensor.h"
+#include "paddle/framework/operator.h"
+
+namespace paddle {
+namespace operators {
+
+/*
+ * This is an implementation of beam search algorithm, input the candidate ids
+ * and candidate scores, it will sort and select the top beam_size candidates,
+ * return the corresponding ids and scores.
+ */
+class BeamSearchAlgorithm {
+ public:
+  // TODO(superjom) make type customizable
+  using id_t = size_t;
+  using score_t = float;
+  /*
+   * Input the arguments that needed by this class.
+   */
+  BeamSearchAlgorithm(const framework::LoDTensor& ids,
+                      const framework::LoDTensor& scores, size_t level,
+                      size_t beam_size)
+      : beam_size_(beam_size),
+        ids_(&ids),
+        scores_(&scores),
+        lod_level_(level) {}
+
+  /*
+   * The main function of beam search.
+   *
+   * @selected_ids: a [None, 1]-shaped tensor with LoD.
+   *   In a machine translation model, it might be the candidate term id sets,
+   *   each set stored as a varience-length sequence.
+   *   The format might be described with a two-level LoD
+   *   - [[0 1]
+   *   -  [0 1 2]]
+   *   - [[]
+   *   -  [0 1]]
+   *   the first level of LoD tells that there are two source sentences. The
+   *   second level describes the details of the candidate id set's offsets in
+   * the
+   *   source sentences.
+   *
+   *  @selected_scores: a LoD tensor with the same shape and LoD with
+   * selected_ids.
+   *   It stores the corresponding scores of candidate ids in selected_ids.
+   *
+   * Return false if all the input tensor is empty, in machine translation task
+   * that means no candidates is provided, and the task will stop running.
+   */
+  bool operator()(framework::LoDTensor* selected_ids,
+                  framework::LoDTensor* selected_scores);
+
+ protected:
+  /*
+   * The basic items help to sort.
+   */
+  struct Item {
+    Item() {}
+    Item(size_t offset, size_t id, float score)
+        : offset(offset), id(id), score(score) {}
+    // offset in the lod_level_+1
+    size_t offset;
+    // the candidate id
+    id_t id;
+    // the corresponding score
+    score_t score;
+  };
+
+  /*
+   * Collect the candidate ids that is selected by beam search and store them
+   * into a LoDTensor.
+   */
+  void CollectSelectedResult(framework::LoDTensor* selected_ids,
+                             framework::LoDTensor* selected_scores);
+
+  /*
+   * Transform the items into a map whose key is offset, value is the items.
+   * NOTE low performance
+   */
+  std::map<size_t, std::vector<Item>> CollectItems(
+      const std::vector<std::vector<Item>>& inputs);
+
+  /*
+   * For each source, select top beam_size records.
```

Comment:
For each unfinished prefix. I think we can avoid using "source" because it is not a common concept in standard beam search.